---
title: "A tale of two genomes: Phylogenetic analyses of contentious relationships with two *Mixta* species (*Erwiniaceae*) and identification of horizontal gene transfer events"
output:
  word_document:
    reference_docx: Styles1.docx
    fig_caption: true
    toc: true
    toc_depth: 5
# bibliography: References.bib
# csl: nature-genetics.csl
---

<!-- A Thesis -->

<!-- Submitted to the Department of Biology -->

<!-- In Partial Fulfillment of the Requirements -->

<!-- For the Degree of -->

<!-- <br> -->

<!-- Bachelor of Science -->

<!-- in -->

<!-- Honours Biology -->

<!-- University of Regina -->

<!-- <br> -->

<!-- By -->

<!-- Kimberly Hinz -->

<!-- Regina, Saskatchewan -->

<!-- April, 2020   -->

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```
```{=openxml}
<w:p><w:r><w:br w:type="page"/></w:r></w:p>
```
# Abstract

This is the abstract for my honours thesis.

<!-- ```{=openxml} -->
<!-- <w:p><w:r><w:br w:type="page"/></w:r></w:p> -->
<!-- ``` -->
<!-- # Acknowledgements -->

<!-- I would like to thank Dr. John Stavrinides for the opportunity of working in his lab and for being incredibly encouraging, supportive, energetic, and motivating as my Honours supervisor. I could not have asked for a better supervisor and experience. -->

<!-- Thank you to Dr. Maria Davis and Dr. Chris Somers for devoting time out of their busy schedules to serve as my committee members. -->

<!-- A big thank you to the Stavrinauts (Stav Lab; SLab) for all their help and support, especially when none of us knew what my results showed. You taught me that research is about finding answers together, not having them beforehand. I would especially like to thank Craig Soutar, Michi Sawa, and Emily Haidl for their knowledge and expertise with bioinformatics and R coding. -->

<!-- Another big thank you to Stefano Mezzini, who believed in my code and R Markdown, even when I did not. And who always found ridiculously simple solutions to my problems. -->

<!-- Lastly, thank you so much to my family and friends for all of your support, even when you did/do not understand what I study. -->

<!-- ```{=openxml} -->
<!-- <w:p><w:r><w:br w:type="page"/></w:r></w:p> -->
<!-- ``` -->
<!-- # Dedication -->

<!-- To my parents, who have loved and supported me throughout all my ventures: the ups, the downs, and the flat lands. Thank you for everything you have done and sacrificed for me. I love you so much! -->

<!-- To Christina Solinas, who sparked my passion for genetics all those years ago. I may not have realized it back then, but biology was what I was always meant to study. Thank you; you have changed my life. -->

<!-- ```{=openxml} -->
<!-- <w:p><w:r><w:br w:type="page"/></w:r></w:p> -->
<!-- ``` -->
# Table of Contents


<!-- ```{=openxml} -->
<!-- <w:p><w:r><w:br w:type="page"/></w:r></w:p> -->
<!-- ``` -->
# List of Figures


<!-- ```{=openxml} -->
<!-- <w:p><w:r><w:br w:type="page"/></w:r></w:p> -->
<!-- ``` -->
# List of Abbreviations
```{r echo = FALSE, message = FALSE, fig.cap=TRUE}
library("dplyr")
abbrevs <- as.data.frame(matrix(nrow = 3, ncol = 0))
abbrevs <- mutate(abbrevs, 
                  Abbreviation = c("GTR", "HKY", "NCBI"),
                  `Full Name` = c("General Time Reversible","Hasegawa-Kishino-Yano",  "National Centre for Biotechnology Information"),)


library("knitr")
kable(abbrevs)
```

<!-- ```{=openxml} -->
<!-- <w:p><w:r><w:br w:type="page"/></w:r></w:p> -->
<!-- ``` -->
# List of Appendices


```{=openxml}
<w:p><w:r><w:br w:type="page"/></w:r></w:p>
```
# Introduction



# Materials and Methods
## Genomes and Homologous Genes
Publicly available full genomes were retrieved from the National Centre for Biotechnology Information (NCBI) comprising of ten type strains from *Mixta*, *Pantoae*, *Erwinia*, *Tatumella*, *Citrobacter*, and *Enterobacter*. A list of these strains follows:

```{r echo = FALSE, message = FALSE, fig.cap=TRUE}
library("dplyr")
strain <- as.data.frame(matrix(nrow = 10, ncol = 0))
strain <- mutate(strain, 
                 Species = c("Mixta calida", "Mixta gaviniae", "Pantoea agglomerans", "Pantoea septica", "Erwinia amylovora", 
                             "Erwinia tasmaniensis", "Tatumella ptyseos", "Tatumella saanichensis", "Citrobacter freundii", 
                             "Enterobacter cloacae subsp cloacae"),
                 Strain = c("DSM_22759", "DSM_22758", "NBRC_102470", "LMG_5345", "CFBP_1232", "ET1/99", "NCTC_11468", 
                            "NML_06-3099", "NCTC_9750", "ATCC_13047"), 
                 Level = c("complete genome", "complete genome", "contigs", "contigs", "contigs", "complete genome", 
                           "complete genome", "contigs", "complete genome", "complete genome")
)


library("knitr")
kable(strain, caption = "Table __. Species' names, strains, and assembly levels that are involved in the study.")
```

The genomes were annotated with Prokka v1.14.1 [@Seemann2014] using default parameters and homologous genes were extracted using GET_HOMOLOGUES software package [@Contreras-Moreira2013] with the bidirectional best-hit search algorithm using default parameters. GET_HOMOLOGUES extracted 1284 genes that were common between the ten species. 

Within `R`, a filtering function was created to pass forward fasta files with ten, non-truncated sequences. The minimum gene length to pass through the filter was 90% of the length of the longest sequence in that file. This filter ensured that files containing split, missing, or paralogous genes were not included in the analysis. As a result, 1035 files of the 1284 were passed through. 

## Gene Alignment
Gene alignments were done in `R` with a few software packages. The fasta files were read in using the `readDNAStringSet()` function [@Aboyoun2018] (package: `Biostrings`). The alignments were done using the ClustalW algorithm with the `msaClustalW()` [@Bodenhofer2015] (package: `msa`). The number of maximum iterations was set to 100 and the other parameters were kept at default settings [@Bodenhofer2015]. Alignments were then converted into a readable format using `msaConvert` [@Bodenhofer2015] (package: `msa`) and then written as fasta files using `write.fasta()` [@Jordan2007] (package: `seqinr`).

## Distance Matrices
Prior to computing distance matrices, model testing was done for each of the 1035 gene files through MEGA-CC [@Kumar2012] by calling a text file containing the file pathways and a MAO file detailing the analysis preferences. These results were exported as comma-separated values (CSV) and read into `R`. The best model was chosen based on the lowest BIC and if it was available for distance matrices in MEGAX [@Kumar2012]. General Time Reversible (GTR) models and Hasegawa-Kishino-Yano (HKY) models were not available for distance matrices and so these models were not considered during model testing. Furthermore, the rates among sites option invariant sites (+I) was also not available for distance matrices and were subsequently ignored.

The gene files were separated into groups according to the suggested model and text files were created with their file pathways. These text files were called in turn along with a corresponding MAO file through MEGA-CC to compute the distance matrices for each gene. Variance was estimated with 500 bootstraps, transitions and transversions were allowed, and gaps and missing data were treated with pairwise deletion.

The genetic distances and standard errors relative to *M. calida* and *M. gaviniae* were extracted to create plots in `R`, categorize the closest relatives, run generalized additive models to determine potential non-linear relationships between parameters and genetic distances, and identify genes for further analysis using phylogenetic trees. The phylogenetic trees were created using MEGAx and with a model suggested through model testing.

All `R` code written for the analysis of the 10 genomes, homologous genes, aligned files, distance matrices, and other supplementary results are available on the Github repository associated with this paper (LINK!!!)

# Results and Discussion


# Conclusion

# Tips

For hyperlinks: <http://rmarkdown.rstudio.com>

For bold: **Knit**

For italics: *Italy*

For code: `echo = FALSE`

<!-- ```{=openxml} -->
<!-- <w:p><w:r><w:br w:type="page"/></w:r></w:p> -->
<!-- ``` -->
<!-- # Literature Cited -->


<!-- # Appendix A -->