---
title: "A tale of two genomes: Analyses of contentious phylogenetic relationships of two *Mixta* species (*Erwiniaceae*) and identification of horizontal gene transfer events"
output:
  word_document:
    reference_docx: Styles1.docx
    fig_caption: true
    toc: true
    toc_depth: 5
bibliography: References.bib
csl: nature-genetics.csl
---
# Beginning
A Thesis

Submitted to the Department of Biology

In Partial Fulfillment of the Requirements

For the Degree of

<br>

Bachelor of Science

in

Honours Biology

University of Regina

<br>

By

Kimberly Hinz

Regina, Saskatchewan

April, 2020

```{=openxml}
<w:p><w:r><w:br w:type="page"/></w:r></w:p>
```
# Abstract

*Mixta* is a recently described genus within the bacterial family *Erwiniaceae* that is closely related to *Pantoea*, *Erwinia*, and *Tatumella*. Depending on the genes and method used for phylogenetic analysis, the genus with which *Mixta* shares the most recent ancestor differs. This study aimed to determine the cause behind these contentious results, and identified genes that resulted from horizontal gene transfer. Ten complete or partial genomes – two *Mixta* species, six species from other *Erwiniaceae* genera, and two outgroups – were retrieved from NCBI and annotated. Homologous genes were extracted yielding a dataset of 1035 genes. The genes were aligned With the ClustalW algorithm and MEGA-CC was used for model testing and computing the distance matrix for each gene.

**Results**

**Conclusions:** 

# Acknowledgements/Dedication
<!-- ```{=openxml} -->
<!-- <w:p><w:r><w:br w:type="page"/></w:r></w:p> -->
<!-- ``` -->
<!-- # Acknowledgements -->

<!-- I would like to thank Dr. John Stavrinides for the opportunity of working in his lab and for being incredibly encouraging, supportive, energetic, and motivating as my Honours supervisor. I could not have asked for a better supervisor and experience. -->

<!-- Thank you to Dr. Maria Davis and Dr. Chris Somers for devoting time out of their busy schedules to serve as my committee members. -->

<!-- A big thank you to the Stavrinauts (Stav Lab; SLab) for all their help and support, especially when none of us knew what my results showed. You taught me that research is about finding answers together, not having them beforehand. I would especially like to thank Craig Soutar, Michiko Sawa, and Emily Haidl for their knowledge and expertise with bioinformatics and R coding. -->

<!-- Another big thank you to Stefano Mezzini, who believed in my code and R Markdown, even when I did not and who always found ridiculously simple solutions to my problems. And thank you for being my friend, you have made this journey so much easier and fun. -->

<!-- Lastly, thank you so much to my family and friends for all of your support, even when you did/do not understand what I study. -->

<!-- ```{=openxml} -->
<!-- <w:p><w:r><w:br w:type="page"/></w:r></w:p> -->
<!-- ``` -->
<!-- # Dedication -->

<!-- To my parents, who have loved and supported me throughout all my ventures: the ups, the downs, and the flat lands. Thank you for everything you have done and sacrificed for me. I love you so much! -->

<!-- To Christina Solinas, who sparked my passion for genetics all those years ago. I may not have realized it back then, but biology was what I was always meant to study. Thank you; you have changed my life. -->

```{=openxml}
<w:p><w:r><w:br w:type="page"/></w:r></w:p>
```
# Table of Contents


```{=openxml}
<w:p><w:r><w:br w:type="page"/></w:r></w:p>
```
# List of Figures


```{=openxml}
<w:p><w:r><w:br w:type="page"/></w:r></w:p>
```
# List of Abbreviations
```{r echo = FALSE, message = FALSE, fig.cap=FALSE}
abbrevs <- data.frame(Abbreviation = c("bp", "GAM", "GTR", "HKY", "HGT", "MLSA", "NCBI", "rRNA"),
                  `Full Name` = c("Base pairs", "Generalized additive model", "General Time Reversible",
                                  "Hasegawa-Kishino-Yano", "Horizontal gene transfer", 
                                  "Multi-locus sequence analysis",
                                  "National Center for Biotechnology Information", "ribosomal RNA"))

library("knitr")
kable(abbrevs)
```

<!-- ```{=openxml} -->
<!-- <w:p><w:r><w:br w:type="page"/></w:r></w:p> -->
<!-- ``` -->
# List of Appendices


```{=openxml}
<w:p><w:r><w:br w:type="page"/></w:r></w:p>
```
# Introduction
<!-- The bacterial family *Erwiniaceae* includes diverse Gram-negative *[@Winslow1917]*\ *[@Prakash2015]* organisms including the genera *Mixta*, *Pantoea*, *Erwinia*, *Tatumella*, Buchnera, Wigglesworthia, Phaseolibacter, Kalamiella (LPSN, http://www.bacterio.net/). Three of these, *Mixta*, *Pantoea*, and *Erwinia*, have been the focus of numerous studies given their relevance to plant and human pathogens. For example, the genus *Erwinia* contains plant pathogens (e.g. *E. amylovora*) *[@HaubenSwings2005]*\ *[@Lopez2011]*\ *[@Matsuura2012]*, epiphytes (e.g. *E. tasmaniensis*) *[@Geider2006]*\ *[@Mergaert1999]*, species that may contribute to plant disease *[@Buonaurio2015]*, and others that are insect-associated *[@Campillo2015]*\ *[@Harada1997]*\ *[@Skrodenyte-Arbaciauskiene2012]*. Similarly, *Pantoea*, a genus reported by Gavini et al. *[@Gavini1989]*, includes species that inhabit a variety of environments *[@Gavini1989]*\ *[@Popp2010]*\ *[@Liu2013]*, including human and plant pathogens *[@Prakash2015]*\ *[@deBaere2004]*\ *[@Fullerton2007]*\ *[@Kratz2003]*\ *[@Lim2006]*\ *[@Schmid2003]*\ *[@vanRostenberghe2006]* as well as species associated with plant growth *[@Smits2011]*\ *[@Kim2012]*, insects *[@Palmer2016]*, and fungi *[@Ma2016]*\ *[@Rong2016]*. *Mixta*, originally *Pantoea*, is a novel genus first suggested by Palmer et al. *[@Palmer2017]* and then later proposed by the same researchers in 2018 *[@Palmer2018]*. Two species, *M. calida* and *M. gaviniae* (previously species of *Pantoea*), were originally found in infant formula and its production environment *[@Popp2010]* and Fritz et al. *[@Fritz2014]* demonstrated that some strains of *M. calida* were linked with post-surgical meningitis in humans. Currently, there are only five *Mixta* species validly published (LPSN, http://www.bacterio.net/). -->

<!-- Understanding the phylogeny of bacterial species and genes allows researchers to investigate the evolutionary history of the species, their relatedness, and the individual genes that allow the species to survive under various conditions *[@Palmer2017]*\ *[@Bennett2012]*\ *[@Prasanna2013]*\ *[@Angus2014]*\ *[@deMaayer2014]*\ *[@Fouts2016]*. The most common methods in phylogenetic analysis are those including 16S ribosomal RNA (rRNA) sequences *[@Konstantinidis2007]*, multi-locus sequence analysis (MLSA) with several housekeeping genes *[@Konstantinidis2007]*\ *[@Gevers2005]*\ *[@Glaeser2015]*, and ribosomal MLSA with 53 structural ribosomal proteins *[@Palmer2017]*\ *[@Bennett2012]*\ *[@Jolley2012]*. However, the phylogenetic trees generated from these methods may not be very robust *[@Konstantinidis2007]*\ *[@Glaeser2015]*\ *[@Brady2008]* because of the lower phylogenetic signal from the highly conserved housekeeping genes *[@Konstantinidis2007]*\ *[@Gevers2005]*\ *[@Fox1992]*\ *[@Konstantinidis2005]*\ *[@Staley2006]*\ *[@RichterRosselloMora2009]*. This is particularly seen within *Erwiniaceae* wherein most of our understanding of the phylogeny and interrelationships are based on the 16S rRNA genes *[@Francino2006]*\ *[@Hauben1998]*\ *[@Naum2008]*\ *[@Sproer1999]*\ *[@Adeolu2016]* because of its ubiquity. However, it has insufficient resolution at the species level *[@Glaeser2015]* and the genera *Pantoea*, *Erwinia*, *Tatumella*, and *Mixta* are often not monophyletic *[@Rezzonico2016]*. Another issue that may arise with MLSA is unknowingly including paralogues (i.e. genes originating from an intragenomic duplication event) or xenologues (i.e. genes originating from horizontal gene transfer) *[@Palmer2017]*\ *[@Koonin2005]. -->

<!-- The DNA topoisomerase (gyrase) subunit (gyrB), DNA-directed RNA polymerase subunit (rpoB), ATP synthase subunit (atpD), and initiation translation factor (infB) have been shown to be more reliable than 16S rRNA gene sequences for determining intra- and intergeneric relationships *[@Hedegaard1999]*\ *[@Dauga2002]*\ *[@Mollet1997]*\ *[@Paradis2005]* and a robust method for the *Erwiniaceae* family *[@Brady2010]*. The genera within *Erwiniaceae* are monophyletic when concatenated partial gyrB, rpoB, atpD, and infB gene sequences are used *[@Brady2008]*\ *[@Brady2009]*\ *[@Moretti2011]*. -->

<!-- Nevertheless, despite efforts to find a suitable method for identifying the species tree for *Erwiniaceae* genera, *Mixta*'s position with relation to *Pantoea*, *Erwinia*, and *Tatumella* vastly differs between studies *[@Prakash2015]*\ *[@Palmer2017]*\ *[@Palmer2018]*\ *[@Adeolu2016]*\ *[@Brady2011]*\ *[@Rezzonico2016]*\ *[@Brady2012]*\ *[@Gueule2015]*. Not only do different gene combinations provide different results, but different algorithms and models for alignments and phylogenetic trees also contribute the contentious results *[@Naum2008]*\ *[@Octavia2014]*. For example, Prakash et al. *[@Prakash2015]*, Palmer et al. *[@Palmer2017]*, Rezzonico et al. *[@Rezzonico2016]*, and Gueule et al. *[@Gueule2015]* all used gyrB, rpoB, atpD, and infB nucleotide gene sequences in their phylogenetic analyses. Prakash et al. conducted a neighbour-joining statistical method while the others used a maximum-likelihood method. Three of the studies – Rezzonico et al., Prakash et al., and Gueule et al. – placed the *Mixta* species within the *Pantoea* genera whereas the analysis in Palmer et al. places *Mixta* outside of the other three genera. Palmer et al. *[@Palmer2018]*, Rezzonico et al. *[@Rezzonico2016]*, and Brady et al. *[@Brady2012]* also ran analyses with the four MLSA genes as amino acid sequences. These three studies show contentious results as Rezzonico et al. shows that *Mixta* is related closest with *Tatumella*, Palmer et al. shows that *Mixta* is related closest with *Pantoea*, and Brady et al. shows *Mixta* within the *Pantoea* genus. In a separate study conducted by Palmer et al. *[@Palmer2017]*, two maximum-likelihood trees were computed, one with 1039 core amino acid sequence and the other with 52 ribosomal MLSA nucleotide genes. Both maximum-likelihood trees contained species from *Mixta*, *Pantoea*, *Erwinia*, and *Tatumella*; however, the analysis with the core genes placed *Pantoea* and *Tatumella* as sister taxa followed by two *Mixta* species whereas the analysis with ribosomal MLSA genes placed *Pantoea* and *Tatumella* as sister taxa followed by *Erwinia* and then by the *Mixta* species. A possible cause behind these contentious results may be horizontal gene transfer (HGT) events within the *Erwiniaceae* or from unrelated species. HGT events occur frequently within bacteria and often the genes provide the bacteria the ability to adapt to new environments and/or survive selective pressures. -->

<!-- The objective of my honours thesis was to determine the cause behind the contentious phylogenetic results for *Mixta* species. To accomplish this, ten complete or partial genomes – two each for *Mixta*, *Pantoea*, *Erwinia*, *Tatumella*, and outgroup – were retrieved from public databases, annotated, and homologous genes were extracted. Genes were used to make individual phylogenetic trees, and the trees were examined for expected and unusual relative patterns. -->

# Materials and Methods
<!-- ## Genomes and Gene Sets -->
<!-- Publicly available full genomes were retrieved from NCBI comprising of ten type strains from *Mixta*, *Pantoae*, *Erwinia*, *Tatumella*, *Citrobacter*, and *Enterobacter* (Table 1). The genomes were annotated with Prokka v1.14.1 *[@Seemann2014]* using default parameters and homologous genes were extracted using GET_HOMOLOGUES software package *[@Contreras-Moreira2013]* with the bidirectional best-hit search algorithm using default parameters. GET_HOMOLOGUES extracted 1284 genes that were common between the ten species. -->

<!-- ```{r TABLE1, echo = FALSE, message = FALSE, fig.cap=TRUE} -->
<!-- strain <- data.frame(Species = c("Mixta calida", "Mixta gaviniae", "Pantoea agglomerans", -->
<!--                                  "Pantoea septica", "Erwinia amylovora", "Erwinia tasmaniensis", -->
<!--                                  "Tatumella ptyseos", "Tatumella saanichensis", "Citrobacter freundii", -->
<!--                                  "Enterobacter cloacae subsp cloacae"), -->
<!--                      Strain = c("DSM_22759", "DSM_22758", "NBRC_102470", "LMG_5345", "CFBP_1232", -->
<!--                                 "ET1/99", "NCTC_11468", "NML_06-3099", "NCTC_9750", "ATCC_13047"), -->
<!--                      GenBank = c("GCA_002953215.1", "GCA_002953195.1", "GCA_001598475.1", -->
<!--                                  "GCA_002095575.1", "GCA_000367625.2", "GCA_000026185.1", -->
<!--                                  "GCA_900478715.1", "GCA_000439375.1", "GCA_011064845.1", -->
<!--                                  "GCA_000025565.1"), -->
<!--                      Level = c("complete", "complete", "contigs", "contigs", "contigs", "complete", -->
<!--                                "complete", "contigs", "complete", "complete")) -->
<!-- colnames(strain)[3] <- "GenBank Accession Number" -->

<!-- library("knitr") -->
<!-- kable(strain, caption = "Table 1. Genomes used in the study.") -->
<!-- ``` -->

<!-- Within `R` *[@RCore2019]*, a filtering function was created to pass forward fasta files with ten, non-truncated sequences. The minimum gene length to pass through the filter was 90% of the length of the longest sequence in that file. This filter ensured that files containing split, missing, or paralogous genes were not included in the analysis. As a result, 1035 files of the 1284 passed this filtering stage. -->

<!-- ## Gene Alignment -->
<!-- Gene alignments were done in `R` with a few software packages. The fasta files were read in using the `readDNAStringSet()` function *[@Aboyoun2018]* (package: `Biostrings`). The alignments were made using the ClustalW algorithm with the `msaClustalW()` *[@Bodenhofer2015]* (package: `msa`). The number of maximum iterations was set to 100 and the other parameters were kept at default settings *[@Bodenhofer2015]*. Alignments were then converted into a readable format using `msaConvert` *[@Bodenhofer2015]* (package: `msa`) and then written as fasta files using `write.fasta()` *[@Jordan2007]* (package: `seqinr`). -->

<!-- ## Distance Matrices -->
<!-- Prior to computing distance matrices, model testing was done for each of the 1035 gene files through MEGA-CC *[@Kumar2012]* by calling a text file containing the file pathways and a MAO file detailing the analysis preferences. These results were exported as comma-separated values (CSV) and read into `R`. The best model was chosen based on the lowest Bayesian Information Criterion (BIC) and if it was available for distance matrices in MEGAX *[@Kumar2018]*. General Time Reversible (GTR) models and Hasegawa-Kishino-Yano (HKY) models were not available for distance matrices and so these models were not considered during model testing. Furthermore, the rates among sites option invariant sites (+I) was also not available for distance matrices and were subsequently ignored. -->

<!-- The gene files were separated into groups according to the suggested model and text files were created with their file pathways. A table containing the genes included in this study and the model used for each gene is available in **Supplementary Table 1 on Github**. These text files were called in turn along with a corresponding MAO file through MEGA-CC to compute the distance matrices for each gene. Variance was estimated with 500 bootstraps, transitions and transversions were allowed, and gaps and missing data were treated with pairwise deletion. -->

<!-- The genetic distances and standard errors relative to *M. calida* and *M. gaviniae* were extracted to create plots in `R`, categorize the closest relatives, run generalized additive models to determine potential non-linear relationships between parameters and genetic distances, and identify genes for further analysis using phylogenetic trees. The maximum-likelihood phylogenetic trees were created using MEGAX and with a model suggested through model testing. -->

# Results and Discussion
## Generalized Additive Models
Generalized additive models (GAMs) were fit to determine if gene length affected evolutionary distance from or the closest relative to the two *Mixta* species. The effect of gene length on the evolutionary distances is shown in Figure 1. Genes with less than 300 base pairs (bp) have a negative effect so all species in the study clustered closely with *Mixta*. This is due to the very low phylogenetic signal of these genes. The effect is also negative for gene lengths between 2750 to 3250 bp and longer than 4000; however, this is due to lack of genes in the dataset with these lengths. This is also the case for gene lengths between 3500 and 3900 bp where there is a positive effect.

```{r FIGURE1, echo = FALSE, message = FALSE, warning = FALSE, fig.cap = "Figure 1. The effect of gene length on evolutionary distances from (a) *M. calida* and (b) *M. gaviniae*. Zero is the average distance. Grey ribbon is the 95% confidence interval. Plots were made using the `draw()` function (package: `gratia`).", fig.height = 3.25, fig.width = 6.5, fig.align = "center"}
# library("knitr")
# library("gratia")
# library("cowplot")
# library("ggplot2")
# theme_set(theme_bw())
# 
# D_GLc <- readRDS("C:/Users/officePC/Documents/Kim_Honours/Mixta_Mosaic/10Gam_Models/MC_Distance.rds")
# D_GLg <- readRDS("C:/Users/officePC/Documents/Kim_Honours/Mixta_Mosaic/10Gam_Models/MG_Distance.rds")
# 
# cal <- draw(D_GLc)
# gav <- draw(D_GLg)
# 
# plt <- plot_grid(cal, gav, ncol = 2, labels = "auto", align = "hv", axis = "tblr"); plt
```

The second GAM modelled how the probability of the first non-*Mixta* relative to the *Mixta* species being a given species changed over the range of mean gene lengths (Figure 2). The closest relative is the species with the shortest evolutionary history from the *Mixta* species. Throughout the range of gene lengths, *P. septica* (light blue line) is the most likely species to be closely related to both *M. calida* and *M. gaviniae*. This suggests that the *Mixta* species are most closely related to *P. septica* and that most genes will show this pattern.

```{r FIGURE2, echo = FALSE, message = FALSE, warning = FALSE, fig.cap = "Figure 2. The probability of the first relative to (a) *M. calida* and (b) *M. gaviniae* being any of the species in the study given mean gene length. Grey ribbons are the 95% confidence intervals. Plots were made using the `ggplot` function (package: `ggplot2`).", fig.height = 6.5, fig.width = 6.5, fig.align = "center"}
# library("knitr")
# library("dplyr")
# library("gratia")
# library("cowplot")
# library("ggplot2")
# theme_set(theme_bw())
# 
# M_cal_pred <- read.csv(file = "C:/Users/officePC/Documents/Kim_Honours/Mixta_Mosaic/For_Paper/GAM_predictions_calida.csv",
#                        stringsAsFactors = FALSE)
# M_cal_pred <- mutate(M_cal_pred,
#                      lower_0 = case_when(lower > 0 ~ lower, lower < 0 ~ 0),
#                      upper_1 = case_when(upper < 1 ~ upper, upper > 1 ~ 1))
# 
# MC_pred <- ggplot(M_cal_pred, aes(x = Mean, y = Fit, colour = Species)) +
#   geom_ribbon(aes(ymin = lower_0, ymax = upper_1, x = Mean), inherit.aes = FALSE,
#               data = M_cal_pred, alpha = 0.2) +
#   geom_line() +
#   scale_colour_manual(values = alpha(c("red", "orange", "darkgreen", "green3", "blue3", "dodgerblue2", "darkorchid", "violetred1"))) +
#   theme(text = element_text(size = 9),
#         legend.title = element_blank(),
#         legend.position = "none") +
#   labs(x = "Gene Length Mean", y = "Probability",
#        colour = "Species") +
#   scale_y_continuous(breaks = c(0, 0.25, 0.5, 0.75, 1),
#                      limits = c(-0.01, 1.01), expand = c(0, 0))
# 
# M_gav_pred <- read.csv(file = "C:/Users/officePC/Documents/Kim_Honours/Mixta_Mosaic/For_Paper/GAM_predictions_gaviniae.csv",
#                        stringsAsFactors = FALSE)
# M_gav_pred <- mutate(M_gav_pred,
#                      lower_0 = case_when(lower > 0 ~ lower, lower < 0 ~ 0),
#                      upper_1 = case_when(upper < 1 ~ upper, upper > 1 ~ 1))
# 
# MG_pred <- ggplot(M_gav_pred, aes(x = Mean, y = Fit, colour = Species)) +
#   geom_ribbon(aes(ymin = lower_0, ymax = upper_1, x = Mean), inherit.aes = FALSE,
#               data = M_gav_pred, alpha = 0.2) +
#   geom_line() +
#   scale_colour_manual(values = alpha(c("red", "orange", "darkgreen", "green3", "blue3", "dodgerblue2", "darkorchid", "violetred1")),
#                       labels = c("C. freundii", "E. cloacae", "E. amylovora", "E. tasmaniensis",
#                                  "P. agglomerans", "P. septica", "T. ptyseos", "T. saanichensis")) +
#   theme(legend.position = "bottom", text = element_text(size = 12),
#         legend.text = element_text(face = "italic")) +
#   labs(x = "Gene Length Mean", y = "Probability",
#        colour = "Species") +
#   scale_y_continuous(breaks = c(0, 0.25, 0.5, 0.75, 1),
#                      limits = c(-0.01, 1.01), expand = c(0, 0))
# 
# plt2 <- plot_grid(MC_pred, MG_pred, ncol = 1, labels = "auto", align = "hv", axis = "lr", rel_heights = c(1, 1.3)); plt2
```

## First Relative
The first relative – the species with the lowest genetic distance – to both *Mixta* species, exluding each other, was extracted from the distance matrices for each gene. The total number and percentage of genes for each species is given in Table 2. *P. septica* was the closest relative for 80.0% and 81.0% of the genes for *M. calida* and *M. gaviniae*, respectively. This further supports that *P. septica* is the closest relative to *Mixta* from the species in the analysis followed by *P. agglomerans*.

```{r TABLE1, echo = FALSE, message = FALSE, fig.cap = TRUE}
library("knitr")
library("dplyr")
library("gratia")
library("cowplot")
library("ggplot2")
theme_set(theme_bw())

M_cal <- read.csv(file = "C:/Users/officePC/Documents/Kim_Honours/Mixta_Mosaic/For_Paper/Plots/MC_circle_data.csv",
                  stringsAsFactors = FALSE)
M_cal <- subset(M_cal, Results_Number != "NA")
M_cal <- mutate(M_cal,
                Species = case_when(Results_Other == "Pantoea_agglomerans" ~ "P. agglomerans", Results_Other == "Pantoea_septica" ~ "P. septica",
                                    Results_Other == "Erwinia_amylovora" ~ "E. amylovora", Results_Other == "Erwinia_tasmaniensis" ~ "E. tasmaniensis",
                                    Results_Other == "Tatumella_ptyseos" ~ "T. ptyseos", Results_Other == "Tatumella_saanichensis" ~ "T. saanichensis",
                                    Results_Other == "Citrobacter_freundii" ~ "C. freundii", Results_Other == "Enterobacter_cloacae" ~ "E. cloacae"))

spcs <- data.frame(Species = c("P. septica", "P. agglomerans", "E. tasmaniensis", "E. amylovora", "T. ptyseos", "T. saanichensis", "C. freundii", "E. cloacae"))

cal <- as.data.frame(matrix(nrow = 0, ncol = 0))
for(row in 1:nrow(spcs)) {
  numspec <- data.frame(Species = spcs$Species[row],
                        Number = length(which(M_cal$Species == spcs$Species[row])))
  
  cal <- rbind(cal, numspec)
}
rm(numspec, row)

M_gav <- read.csv(file = "C:/Users/officePC/Documents/Kim_Honours/Mixta_Mosaic/For_Paper/Plots/MG_circle_data.csv",
                  stringsAsFactors = FALSE)
M_gav <- subset(M_gav, Results_Number != "NA")
M_gav <- mutate(M_gav,
                Species = case_when(Results_Other == "Pantoea_agglomerans" ~ "P. agglomerans", Results_Other == "Pantoea_septica" ~ "P. septica",
                                    Results_Other == "Erwinia_amylovora" ~ "E. amylovora", Results_Other == "Erwinia_tasmaniensis" ~ "E. tasmaniensis",
                                    Results_Other == "Tatumella_ptyseos" ~ "T. ptyseos", Results_Other == "Tatumella_saanichensis" ~ "T. saanichensis",
                                    Results_Other == "Citrobacter_freundii" ~ "C. freundii", Results_Other == "Enterobacter_cloacae" ~ "E. cloacae"))

gav <- as.data.frame(matrix(nrow = 0, ncol = 0))
for(row in 1:nrow(spcs)) {
  numspec <- data.frame(Species = spcs$Species[row],
                        Number = length(which(M_gav$Species == spcs$Species[row])))
  
  gav <- rbind(gav, numspec)
}
rm(numspec, row)

cal$Percent <- round((cal$Number / 1035) * 100, digits = 1)
cal$numper <- paste(cal$Number, " (", cal$Percent, "%)", sep = "")
gav$Percent <- round((gav$Number / 1035) * 100, digits = 1)
gav$numper <- paste(gav$Number, " (", gav$Percent, "%)", sep = "")

Num_Perc <- data.frame(Species = cal$Species,
                       calida = cal$numper,
                       gaviniae = gav$numper)

colnames(Num_Perc) <- c("Species", "M. calida", "M. gaviniae")


kable(Num_Perc, caption = "Table 2. Number and percentages of genes with each species as the first relative to the Mixta species.")
```

```{r FIGURE3, echo = FALSE, message = FALSE, warning = FALSE, fig.cap = "Figure 3. Circular representation of the first relative to *M. calida* across the genome. Numbers are the gene IDs, and colours and lengths of the lines represent the different species.", fig.align = "center"}
library("knitr")
knitr::include_graphics("C:/Users/officePC/Documents/Kim_Honours/Mixta_Mosaic/For_Paper/Plots/MC_categ_results.png")
```

```{r FIGURE4, echo = FALSE, message = FALSE, warning = FALSE, fig.cap = "Figure 4. Circular representation of the first relative to *M. gaviniae* across the genome. Numbers are the gene IDs, and colours and lengths of the lines represent the different species.", fig.align = "center"}
library("knitr")
knitr::include_graphics("C:/Users/officePC/Documents/Kim_Honours/Mixta_Mosaic/For_Paper/Plots/MG_categ_results.png")
```

* In contrast to 16S rRNA genes, protein-coding genes have been said to have a consistent evolution rate and have more phylogenetic resolution at the genus and/or species level *[@Glaeser2015]*. Multiple-gene-based phylogenies have been developed to overcome the bias of single-gene analysis *[@Glaeser2015]* and they are more likely to show the “true” phylogenetic relationships *[@Glaeser2015]*. This is because they reduce the possibility of ambiguities *[@Hedegaard1999]* by diluting the signal from potential xenologues. It is recommended that single copy, ubiquitous (within the studied taxa) genes are chosen for this type of analysis *[@Glaeser2015]*. 
* Similar to MLSA is using the shared gene content between all taxa included in the study as these genes form the “core” genome *[@Daubin2002]*\ *[@Coenye2005]*\ *[@LanReeves2000]*. The core genome represents a small subset of the taxa’s whole genome *[@Makarova2006]*\ *[@Lukjancenko2012]* and its genes are recognized as essential for survival *[@Hacker2012]*. Additionally, the genes of the core genome are inherited primary through vertical gene transfer *[@Daubin2002]*\ *[@Coenye2005]*.

# Conclusion

<!-- Future: looking into this as a mosaic genome. -->

<!-- # Tips -->

<!-- For hyperlinks: <http://rmarkdown.rstudio.com> -->

<!-- For bold: **Knit** -->

<!-- For italics: *Italy* -->

<!-- For code: `echo = FALSE` -->

```{=openxml}
<w:p><w:r><w:br w:type="page"/></w:r></w:p>
```
# Literature Cited


<!-- # Appendix A -->
